odl/trafos/convolution.py

```diff
@@ -0,0 +1,804 @@
+          :math:`\Omega + \Omega_0`.
+
+          For example, if :math:`\Omega = [0, 5]` and
+          :math:`\Omega_0 = [1, 3]`, i.e. :math:`m_0 = 2`, then
+          :math:`\Omega + \Omega_0 = [1, 8]`. However, we choose the
+          shifted interval :math:`\Omega + m_0 = [2, 7]` for the range
+          of the convolution since it contains the "main mass" of the
+          result.
+        """
+        # Basic checks
+        if not isinstance(domain, DiscreteLp):
+            raise TypeError('domain {!r} is not a DiscreteLp instance'
+                            ''.format(domain))
+        if not domain.is_uniform:
+            raise ValueError('irregular sampling not supported')
+
+        if range is not None:
+            # TODO
+            raise NotImplementedError('custom range not implemented')
+        else:
+            kernel_kwargs = kwargs.pop('kernel_kwargs', {})
+            self._kernel, range = self._compute_kernel_and_range(
+                kernel, domain, kernel_kwargs)
+
+        super().__init__(domain, range, linear=True)
+
+        # Handle impl
+        impl, impl_in = str(impl).lower(), impl
+        if impl not in _REAL_CONV_SUPPORTED_IMPL:
+            raise ValueError("implementation '{}' not understood"
+                             ''.format(impl_in))
+        self._impl = impl
+
+        # Handle the `resample` input parameter
+        resample = normalized_scalar_param_list(kwargs.pop('resample', 'down'),
+                                                length=self.domain.ndim,
+                                                param_conv=str)
+        for i, r in enumerate(resample):
+            if r.lower() not in ('up', 'down'):
+                raise ValueError("in axis {}: `resample` '{}' not understood"
+                                 ''.format(i, r))
+            else:
+                resample[i] = r.lower()
+        self._resample = resample
+
+        # Initialize resampling operators
+        new_kernel_csides, new_domain_csides = self._get_resamp_csides(
+            self.resample, self._kernel.space.cell_sides,
+            self.domain.cell_sides)
+
+        domain_resamp_space = uniform_discr_fromdiscr(
+            self.domain, cell_sides=new_domain_csides)
+        self._domain_resampling_op = Resampling(self.domain,
+                                                domain_resamp_space)
+
+        kernel_resamp_space = uniform_discr_fromdiscr(
+            self._kernel.space, cell_sides=new_kernel_csides)
+        self._kernel_resampling_op = Resampling(self._kernel.space,
+                                                kernel_resamp_space)
+
+        # Scale the kernel if desired
+        self._kernel_scaling = np.prod(new_kernel_csides)
+        self._scale = bool(kwargs.pop('scale', True))
+        if self._scale:
+            # Don't modify the input
+            self._kernel = self._kernel * self._kernel_scaling
+
+    @staticmethod
```

Comment:
Static methods should in general be free. Only reason to have static methods is if users need to ask the class something, in general before initialization, e.g.

```python
for impl im Convolution.valid_impls():
    conv = Convolution(impl=impl)
```